server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RestoreAction.java

```diff
@@ -21,87 +21,108 @@
 
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.nio.charset.StandardCharsets;
 import org.apache.commons.io.IOUtils;
 import org.sonar.api.resources.Language;
 import org.sonar.api.resources.Languages;
 import org.sonar.api.server.ws.Request;
 import org.sonar.api.server.ws.Response;
 import org.sonar.api.server.ws.WebService;
 import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.organization.OrganizationDto;
+import org.sonar.db.permission.OrganizationPermission;
 import org.sonar.db.qualityprofile.QualityProfileDto;
 import org.sonar.server.qualityprofile.BulkChangeResult;
 import org.sonar.server.qualityprofile.QProfileBackuper;
+import org.sonar.server.qualityprofile.QProfileRestoreSummary;
+import org.sonar.server.user.UserSession;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.sonarqube.ws.client.component.ComponentsWsParameters.PARAM_ORGANIZATION;
 
 public class RestoreAction implements QProfileWsAction {
 
   private static final String PARAM_BACKUP = "backup";
 
+  private final DbClient dbClient;
   private final QProfileBackuper backuper;
   private final Languages languages;
-  private final QProfileWsSupport qProfileWsSupport;
+  private final UserSession userSession;
+  private final QProfileWsSupport wsSupport;
 
-  public RestoreAction(QProfileBackuper backuper, Languages languages, QProfileWsSupport qProfileWsSupport) {
+  public RestoreAction(DbClient dbClient, QProfileBackuper backuper, Languages languages, UserSession userSession, QProfileWsSupport wsSupport) {
+    this.dbClient = dbClient;
     this.backuper = backuper;
     this.languages = languages;
-    this.qProfileWsSupport = qProfileWsSupport;
+    this.userSession = userSession;
+    this.wsSupport = wsSupport;
   }
 
   @Override
   public void define(WebService.NewController controller) {
-    controller.createAction("restore")
+    WebService.NewAction action = controller.createAction("restore")
       .setSince("5.2")
       .setDescription("Restore a quality profile using an XML file. The restored profile name is taken from the backup file, " +
         "so if a profile with the same name and language already exists, it will be overwritten. " +
         "Require Administer Quality Profiles permission.")
       .setPost(true)
-      .setHandler(this)
-      .createParam(PARAM_BACKUP)
+      .setHandler(this);
+
+    action.createParam(PARAM_BACKUP)
       .setDescription("A profile backup file in XML format, as generated by api/qualityprofiles/backup " +
         "or the former api/profiles/backup.")
       .setRequired(true);
+
+    wsSupport.createOrganizationParam(action)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this instance-reference to a static reference. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2209)